### PR TITLE
[Impeller] Improve atlas blending performance by reducing size of subpass.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Previously, the subpass for draw atlas with a blend mode would be as large as the final coverage needed to perform the draw. This was wasteful for a number of reasons, especially if many of the sample rect + color combinatons were repeated. I also suspect that drawing everything in its final position would lead to different rendering when colors would overlap. Finally, we need to use FramebufferBlendContents on iOS to avoid two subpasses.

This PR attempts to solve all of these problems by compressing (via color + sample rect) the drawAtlas call into a second atlas which has no overlapping and positions the sample rects in a very simple lightly packed atlas, decreasing the size of the subpass. In the wonderous app on iPads, this makes the artifact discovery confetti stuff substantially less jank.

Fixes https://github.com/flutter/flutter/issues/120925
https://github.com/flutter/flutter/issues/118914

### After:

Anywhere from 2x to 3x as fast as before.

image

### CODE REVIEW

File: impeller/entity/contents/atlas_contents.cc [hunk 1]
```diff
@@ -57,6 +64,85 @@ void AtlasContents::SetCullRect(std::optional<Rect> cull_rect) {
   cull_rect_ = cull_rect;
 }
 
+struct AtlasBlenderKey {
+  Color color;
+  Rect rect;
+
+  struct Hash {
+    std::size_t operator()(const AtlasBlenderKey& key) const {
+      return fml::HashCombine(key.color.red, key.color.green, key.color.blue,
```
Comment (jonahwilliams) [inline impeller/entity/contents/atlas_contents.cc:73]: This is probably safe

Comment (jonahwilliams) [reply impeller/entity/contents/atlas_contents.cc:73]: thinking about this more, there could be some inconsistencies with the hash function and equality implementation given that the latter uses nearly equal. Instead i'll keep the int representation of the color around and use that for both the hash and equality.

**Comment by jonahwilliams** [comment]
Could file a bug for this or write a short doc. Performance issue is apparent in wonderous on iPad.

**Comment by jonahwilliams** [comment]
Bug filled

File: impeller/entity/contents/atlas_contents.cc [hunk 2]
```diff
@@ -57,6 +64,96 @@ void AtlasContents::SetCullRect(std::optional<Rect> cull_rect) {
...
+    }
+
+    auto key = it->first;
+    auto transforms = it->second;
+
+    auto new_rect = Rect::MakeXYWH(x_offset, y_offset, key.rect.size.width,
+                                   key.rect.size.height);
+    auto sub_transform = Matrix::MakeTranslation(Vector2(x_offset, y_offset));
+
+    x_offset += (key.rect.size.width + 1.0);
```
Comment (jonahwilliams) [inline impeller/entity/contents/atlas_contents.cc:139]: actually this logic is probably wrong if the sample rect doesn't have an integral width. I should probably just ceil the width here too.

File: impeller/entity/contents/atlas_contents.cc [hunk 3]
```diff
@@ -57,6 +64,96 @@ void AtlasContents::SetCullRect(std::optional<Rect> cull_rect) {
   cull_rect_ = cull_rect;
 }
 
+struct AtlasBlenderKey {
+  Color color;
+  Rect rect;
+  int32_t color_key;
```
Comment (chinmaygarde) [inline impeller/entity/contents/atlas_contents.cc:70]: `uint32_t`?

File: impeller/entity/contents/atlas_contents.cc [hunk 4]
```diff
@@ -57,6 +64,96 @@ void AtlasContents::SetCullRect(std::optional<Rect> cull_rect) {
...
+
+  static int32_t ToColorKey(Color color) {
+    return (((std::lround(color.alpha * 255) & 0xff) << 24) |
```
Comment (chinmaygarde) [inline impeller/entity/contents/atlas_contents.cc:73]: Ah, the `kEhCloseEnough` of the color world. Perhaps we should have an `IColor` in `color.h` that does this for us?

**Review by chinmaygarde — APPROVED** [review]
The optimization makes sense to me and it neatly isolated in the atlas contents. cc @bdero too in case he has thoughts.

**Review by bdero — APPROVED** [review]
LGTM!

File: impeller/entity/contents/atlas_contents.cc (hunk 3 above)
Comment (jonahwilliams) [reply impeller/entity/contents/atlas_contents.cc]: done

**Pull request merged**